File: benchmark/biquad.lua
-- Biquad filter DSP benchmark
-- Process 10 million samples through a lowpass filter

local sin = math.sin
local pi = math.pi

-- Design a 2nd-order lowpass (RBJ cookbook)
local function lpf(fc, fs, q)
    local w0 = 2.0 * pi * fc / fs
    local alpha = sin(w0) / (2.0 * q)
    local cs = math.cos(w0)

    local a0 = 1.0 + alpha
    local inv = 1.0 / a0

    return {
        b0 = ((1.0 - cs) / 2.0) * inv,
        b1 = (1.0 - cs) * inv,
        b2 = ((1.0 - cs) / 2.0) * inv,
        a1 = (-2.0 * cs) * inv,
        a2 = (1.0 - alpha) * inv,
        x1 = 0.0, x2 = 0.0,
        y1 = 0.0, y2 = 0.0
    }
end

local function run()
    local bq = lpf(1000.0, 44100.0, 0.707)

    local n = 10000000
    local phase = 0.0
    local freq = 440.0 / 44100.0
    local last_y = 0.0
    local two_pi = 2.0 * pi

    -- Process a 440 Hz sine wave through the filter
    for i = 1, n do
        local x = sin(phase * two_pi)
        phase = phase + freq
        if phase > 1.0 then phase = phase - 1.0 end

        local y = bq.b0 * x + bq.b1 * bq.x1 + bq.b2 * bq.x2 - bq.a1 * bq.y1 - bq.a2 * bq.y2
        bq.x2 = bq.x1
        bq.x1 = x
        bq.y2 = bq.y1
        bq.y1 = y
        last_y = y
    end
end

local t0 = os.clock()
run()
local elapsed = os.clock() - t0
io.stderr:write(string.format("exec: %.3fs\n", elapsed))

File: benchmark/biquad.c
#include <stdio.h>
#include <math.h>
#include <time.h>

typedef struct {
    float b0, b1, b2;
    float a1, a2;
    float x1, x2;
    float y1, y2;
} Biquad;

static Biquad lpf(float fc, float fs, float q) {
    float w0 = 2.0f * 3.14159265f * fc / fs;
    float alpha = sinf(w0) / (2.0f * q);
    float cs = cosf(w0);

    float a0 = 1.0f + alpha;
    float inv = 1.0f / a0;

    Biquad bq;
    bq.b1 = (1.0f - cs) * inv;
    bq.b0 = bq.b1 / 2.0f;
    bq.b2 = bq.b0;
    bq.a1 = (0.0f - 2.0f * cs) * inv;
    bq.a2 = (1.0f - alpha) * inv;
    bq.x1 = 0.0f;
    bq.x2 = 0.0f;
    bq.y1 = 0.0f;
    bq.y2 = 0.0f;
    return bq;
}

int main(void) {
    Biquad bq = lpf(1000.0f, 44100.0f, 0.707f);

    int n = 10000000;
    float phase = 0.0f;
    float freq = 440.0f / 44100.0f;
    float two_pi = 2.0f * 3.14159265f;

    clock_t t0 = clock();

    for (int i = 0; i < n; i++) {
        float x = sinf(phase * two_pi);
        phase = phase + freq;
        if (phase > 1.0f) { phase = phase - 1.0f; }

        float y = bq.b0 * x + bq.b1 * bq.x1 + bq.b2 * bq.x2
                - bq.a1 * bq.y1 - bq.a2 * bq.y2;
        bq.x2 = bq.x1;
        bq.x1 = x;
        bq.y2 = bq.y1;
        bq.y1 = y;
    }

    double elapsed = (double)(clock() - t0) / CLOCKS_PER_SEC;
    fprintf(stderr, "exec: %.3fs\n", elapsed);

    // Prevent the compiler from optimizing away the loop.
    volatile float sink = bq.y1;
    (void)sink;
    return 0;
}
